From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fredrik=20S=C3=B6derquist?= <fs@opera.com>
Date: Tue, 20 Oct 2026 23:29:17 -0700
Subject: [M156] Limit image request in
 HTMLPlugInElement::CustomStyleForLayoutObject
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Original change's description:
> Limit image request in HTMLPlugInElement::CustomStyleForLayoutObject
>
> When the ComputedStyle is being ensured we shouldn't perform an image
> load.
>
> Fixed: 574101376, 575184164
> Change-Id: I7b678074d4496ef667f360cbab62ce7c459b4785
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8611107
> Reviewed-by: Rune Lillesveen <futhark@chromium.org>
> Commit-Queue: Rune Lillesveen <futhark@chromium.org>
> Auto-Submit: Fredrik Söderquist <fs@opera.com>
> Cr-Commit-Position: refs/heads/main@{#1722229}

(cherry picked from commit f49d058c880110acd4479e6634cc8013e3f672cd)

Bug: 575705071,574101376,575184164,575705444
Change-Id: I7b678074d4496ef667f360cbab62ce7c459b4785
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/8616178
Commit-Queue: Rune Lillesveen <futhark@chromium.org>
Reviewed-by: Rune Lillesveen <futhark@chromium.org>
Auto-Submit: chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com <chrome-cherry-picker@chops-service-accounts.iam.gserviceaccount.com>
Cr-Commit-Position: refs/branch-heads/8078@{#1375}
Cr-Branched-From: c50edc4214d81e4f944c3415d55104a49e46d214-refs/heads/main@{#1706539}

diff --git a/third_party/blink/renderer/core/html/html_object_element_test.cc b/third_party/blink/renderer/core/html/html_object_element_test.cc
index 7a8e47095638c9757eb9c976d340e0ddbe3a6eb5..adb34ca15722abeeac14728eb9e00bfa90e79a20 100644
--- a/third_party/blink/renderer/core/html/html_object_element_test.cc
+++ b/third_party/blink/renderer/core/html/html_object_element_test.cc
@@ -11,6 +11,8 @@
 #include "third_party/blink/renderer/core/frame/local_frame_view.h"
 #include "third_party/blink/renderer/core/html/html_slot_element.h"
 #include "third_party/blink/renderer/core/testing/dummy_page_holder.h"
+#include "third_party/blink/renderer/core/testing/sim/sim_request.h"
+#include "third_party/blink/renderer/core/testing/sim/sim_test.h"
 #include "third_party/blink/renderer/platform/testing/task_environment.h"
 
 namespace blink {
@@ -50,4 +52,25 @@ TEST_F(HTMLObjectElementTest, FallbackRecalcForReattach) {
   EXPECT_TRUE(slot->GetComputedStyle());
 }
 
+class HTMLObjectElementSimTest : public SimTest {};
+
+TEST_F(HTMLObjectElementSimTest, NoLoadOnEnsureComputedStyle) {
+  SimRequest request("https://example.com/test.html", "text/html");
+  LoadURL("https://example.com/test.html");
+  request.Complete(R"HTML(
+     <!doctype html>
+     <div style='display: none'>
+       <object id='obj' data='file.png' type='image/png'></object>
+     </div>
+  )HTML");
+
+  auto* object =
+      To<HTMLObjectElement>(GetDocument().getElementById(AtomicString("obj")));
+  ASSERT_TRUE(object);
+  EXPECT_FALSE(object->GetComputedStyle());
+  object->EnsureComputedStyle();
+  // Pass if no crash (URLLoaderMock DCHECK failure) because of a request being
+  // performed for 'file.png'.
+}
+
 }  // namespace blink
diff --git a/third_party/blink/renderer/core/html/html_plugin_element.cc b/third_party/blink/renderer/core/html/html_plugin_element.cc
index 181c86bd3be8a71853ae57c96a4d3107b5e27cd5..bb5a853d83695f747ac9b3083bcee738dc31ee10 100644
--- a/third_party/blink/renderer/core/html/html_plugin_element.cc
+++ b/third_party/blink/renderer/core/html/html_plugin_element.cc
@@ -35,6 +35,7 @@
 #include "third_party/blink/renderer/bindings/core/v8/to_v8_traits.h"
 #include "third_party/blink/renderer/core/css/css_property_names.h"
 #include "third_party/blink/renderer/core/css/style_change_reason.h"
+#include "third_party/blink/renderer/core/css/style_recalc_context.h"
 #include "third_party/blink/renderer/core/display_lock/display_lock_utilities.h"
 #include "third_party/blink/renderer/core/dom/document.h"
 #include "third_party/blink/renderer/core/dom/events/event.h"
@@ -936,8 +937,9 @@ const ComputedStyle* HTMLPlugInElement::CustomStyleForLayoutObject(
     const StyleRecalcContext& style_recalc_context) {
   const ComputedStyle* style =
       OriginalStyleForLayoutObject(style_recalc_context);
-  if (IsImageType() && !GetLayoutObject() && style &&
-      LayoutObjectIsNeeded(*style)) {
+  // Don't attempt a load if this computed style is for EnsureComputedStyle().
+  if (!style_recalc_context.is_ensuring_style && IsImageType() &&
+      !GetLayoutObject() && style && LayoutObjectIsNeeded(*style)) {
     if (!image_loader_) {
       image_loader_ = MakeGarbageCollected<HTMLImageLoader>(this);
     }
